contributors/design-proposals/scheduling/pod-priority-resourcequota.md

```diff
@@ -0,0 +1,187 @@
+// its consumption.
+type LimitedResource struct {
+        ...
+
+        // MatchScopes is a collection of filters based on priority classes.
+        // If the object in the intercepted request matches these rules,
+        // quota system will ensure that corresponding quota MUST have
+        // priority based Scopes matches the object in request. 
+        //
+        // If MatchScopes has matched on an object, request will be denied if there is no quota
+        // matches priority class based Scopes. In this case, matching priority class based Scopes
+        // will be an additional requirement for any quota to qualified as covering quota.
+        // +optional
+        MatchScopes []string `json:"matchScopes,omitempty"`
+}
+```
+
+### Expected Behavior of ResourceQuota Admission Controller and ResourceQuota System
+`MatchScopes` will be configured in admission controller configuration to apply quota based on priority class names. If `MatchScopes` matches/selects an incoming pod request, request will be **denied if a Covering Quota is missing**. The meaning of Covering Quota is: any quota which has priority class based `Scopes` that matches/selects the pod in the request. 
+
+Please note that this priority class based criteria will be an **additional** criteria that must be satisfied by covering quota.
+
+To more details, please refer to the `Sample user story` sections at the end of this doc.
+
+#### Backward Compatibility
+
+If the priority class related details the requested pod are not matched/selected by any of the filters in admission controller configuration's `MatchScopes`, overall behavior for the pod will be same as it is today where `ResourceQuota` has no awareness of priority. In such a case, request will be allowed if no covering `ResourceQuota` is found.
+
+Couple of other noteworthy details:
+1. If multiple `ResourceQuota` apply to a Pod, the pod must satisfy all of them.
+2. We do not enforce referential integrity across objects. i.e. Creation or updating of ResourceQuota object, scopes of which names a PriorityClass that does not exist, are allowed.
+3. `ResourceQuota` admission controller **should not reject** deletion of `ResourceQuota` objects with reference.
+
+This design also tries to enable flexibility for its configuration. Here are several sample user stories.
+
+#### Sample User Story 1
+**As a cluster admin, I want `cluster-services` priority only apply to `kube-system` namespace , so that I can ensure those critical daemons on each node while normal user's workloads will not disrupt that ability.**
+
+To enforce above policy:
+1. Admin will create admission controller configuration as below:
+```yaml
+apiVersion: apiserver.k8s.io/v1alpha1
+kind: AdmissionConfiguration
+plugins:
+- name: "ResourceQuota"
+  configuration:
+  apiVersion: resourcequota.admission.k8s.io/v1alpha1
+  kind: Configuration
+  limitedResources:
+  - resource: pods
```

Comment:
Oh, these are K8s resources, AKA object types. How could you support Pods' requested resources, for example, GPU, or Memory?

Reply:
Though the question is important but i would say it is a bit out of scope of this proposal. Use case we are trying to solve is pods with different priority classes and then applying a policy to control which pod is allowed to be created within a namespace.

Reply:
We still lack of device API object in Kube, so better to leave this as is.

Reply:
I agree with leaving special devices alone, but how is your proposal applies to CPU and memory which are supported as first class compute resources? I think quota will have very limited use case if it cannot limit the amount of CPU and memory usage.

Reply:
@bsalamat quota is already supported for CPU and memory resources. If admin has created quotas in a namespace for these resources, and if pod is allowed to be created in this namespace depending on the mechanism/configuration in this PR, then these other compute resource quotas will also apply to the pod. Does this make sense?

Reply:
may be i am missing your question. If my answer does not make sense, can you please rephrase your question?

Reply:
@vikaschoudhary16 Could you please update one of the sample user stories or add another one that shows how an admin can limit the amount of memory in a namespace at a particular priority?

Reply:
sure, thanks!

Reply:
an example resource quota is provided in section 2 so if we just add the yaml output from the kubectl command that is probably fine.

for reference, openshift has some end-user documentation around requiring explicit quota to consume a resource here:
https://docs.openshift.org/latest/admin_guide/quota.html#limited-resources-quota

I think @bsalamat request is pertinent to step 2 in this flow and not step 1 based on my reading.